Call the marc record service generator when there is a catkey or previous catkey

peetucket opened the pull request:
fixes #104 

also fix some rails 5 deprecation warnings

...don't forget to do PRs for shared_config updates for sw-indexer environments before deploy

depends on https://github.com/sul-dlss/dor-services-app/issues/73 being completed and deployed first

cbeer commented:
I confess I'm not clear what problem we're trying to solve (and #104 also only describes a solution), but it strike me as a bit strange for an indexer to trigger side-effects like this. Maybe @blalbrit can enlighten me?

LynnMcRae commented:
The 856 generation in the initial release workflow happens only once, which means the information inserted in Symphony will not get refreshed if there is a catkey change, permissions change (as reflected in $z) or collection membership change.  Having the service called any time the item needs to be reindexed for SW means that information will have the same currency and accuracy for the "indexed via Symphony" solr doc as the "indexed directly from DOR" solr doc. In essence, the 856 is how a with-catkey object gets indexed while an original solr doc is how a non-catkey item gets indexed.

I added a reference to the original JIRA ticket https://jirasul.stanford.edu/jira/browse/JUMBO-37 to #104

cbeer commented:
Thanks for the clarification. That definitely seems to me like a separate process, and not something we should bundle into the indexer.

Obvious candidates, to my mind, are:

- as part of the audit script, run on a schedule in some kind of remediation mode, or
- as part of dor-services, which can run this action in a callback whenever the catkey is changes, or
- as part of this repository, if there's essential logic in this repository this action depends on, but not part of the indexing critical path.

From what I understand, in JUMBO-38, the goal can be expressed as "when the catkey changes, make sure symphony gets the update". By implementing a solution here, we're actually saying "when indexing a record into searchworks, and things may have changes, make sure symphony gets an update" which seems more complicated, carries additional service dependencies with it, and becomes more difficult to test and be sure it is functioning (one failure scenario being "I updated the catkey, but it hasn't been reflected in the symphony record yet, because the indexer hasn't run", perhaps another being: "I updated the catkey for a record that hasn't been released (or was released and then retracted?) yet many weeks ago, but the symphony data is wrong when I release the object, because it hadn't been indexed before".

As a technical architecture, this is troubling too, as we now introduce a service dependency from sw-indexer on the dor-services-app, making it more difficult to test and run without the potential for unexpected side-effects.

LynnMcRae commented:
The job of an indexer for SW is to effect the indexing of a new or changed item in the SW index. For DOR items it has a dependency on SOLR; for Symphony based items it has a dependency on a service that talks to Symphony.  They are equivalent in intent and purpose to me, equally critical (indexing via Symphony is not somehow a secondary sideshow), and belong together in a place that has the responsibility of maintaining the accuracy of the index.  Regardless of a solr error or dor-services-app error, we will be informed, I trust, of any failure in indexing.

The issue is not just catkey, it's a set of indexed information that happens to take a different path to get there. To treat one discovery endpoint for selected information from selected records as the responsibility of the repository as a DOR callback, or to handle half the indexing equation as auditing exceptions seems very wrong to me and way more complicated. 

I understand there are layers involved, and if there are distinctions about what happens in base indexer vs SW-indexer or the need for better encapsulation of solr vs symphony pathways then that's a different issue. Architecturally however I feel strongly sw indexing is where this logic belongs.

LynnMcRae commented:
Can you come to our Tuesday standup, 10:30?

cbeer commented on Gemfile:
```diff
@@ -18,6 +18,8 @@ gem 'okcomputer' # for monitoring
 gem 'rubocop', group: [:development, :test]
 gem 'rubocop-rspec', group: [:development, :test]
 
+gem 'rest-client'
```
Can we use `faraday` instead of `rest-client`? We've been trying to migrate off rest-client after it went through a long period without maintenance, and `faraday` is generally more amenable to features like persistent http connections, http stubbing in tests ,etc.

peetucket replied:
yes, changed it to a post via faraday

cbeer commented on lib/sw_indexer_engine.rb:
```diff
@@ -12,12 +12,15 @@ def index(druid, targets = nil)
-      targets[target_key] = false if purl_model(druid).catkey.present?
+      if purl_model(druid).catkey.present?
+        targets[target_key] = false
+        update_marc_record(druid)
```
I think we want to do this outside the `targets` loop, right? If we have multiple targets, I don't think we need to send multiple messages to the 856 service.

peetucket replied:
yes, that makes sense, though i think do to the way this is written to allow skipping a catkey check for target, it isn't as straightforward as doing another check outside the target loop .... i made some changes to at least ensure it only happens once per druid regardless of how many targets the conditional passes for

cbeer commented on spec/sw_indexer_engine_spec.rb:
```diff
@@ -9,15 +9,19 @@
   describe 'index' do
     context 'for targets that should not be skipped checking' do
-      it 'calls solr delete for the index' do
+      it 'calls solr delete for the index and calls dor-services 856 generation call' do
```
Is there also a test if there isn't a catkey that we ought to add this to?

peetucket replied:
added a test and restructured the other existing tests a bit to make it clearer what was being tested

cbeer commented on lib/sw_indexer_engine.rb:
```diff
@@ -32,4 +35,9 @@ def index(druid, targets = nil)
     DiscoveryIndexer::InputXml::Purlxml.new(druid).load
   end
+
+  def update_marc_record(druid)
```
I wonder if this logic ought to move into the model itself? It might be easier to test there, if nothing else.

peetucket replied:
which model are you referring to?  the sw-indexer has no models

cbeer replied:
Ugh. I assumed `#purl_model` was one of ours. Does it make sense to wrap that in a local class?

peetucket replied:
Sure that's possible, but I guess I don't see a lot of pragmatic value in this.  The code calling these methods is only being executed a few lines above and the overall class itself is already pretty short (the new class would be only a few lines).  Based on how the tests are written, we'd have to alter the test setup and some expectations as well.

cbeer commented:
Before merging this, I think we also need to have an answer in place to deal with changing catkey data, or we're likely to produce bad data downstream. I think we'll want to check at least 3 conditions:

- an item without a catkey is given one
- an item with an existing catkey has the catkey changes
- an item with an existing catkey has its metadata updated

With the last 2, we'll also want to be sure that this works even if the data changes multiple times before Symphony processes the update.

LynnMcRae commented:
- an item without a catkey is given one

Existing function? There is a blind solr-delete-by-druid to assure no orphaned DOR-based doc exists before generating the new catkey entry. One implication is that the (presumably redundant) DOR-based entry will wink out of SW immediately, to be replaced overnight by the PURL in the Symphony catalog record.

- an item with an existing catkey has the catkey changes

New functionality as catkey changes were not automatically propagated. The 856 service will propagate the PURL to the new catkey where it belongs. A controlled catkey change in Argo is being added to record previous catkeys. Any previous_catkey will generate an empty 856 tag for that catkey, causing it to be withdrawn from the previous Symphony record overnight. 

There can be one or (infrequently/rarely) more former catkeys, all which will generate a delete. Identified issue with repeating these deletes over time. Objects tend to be stable once released but need to monitor for how often this happens and do appropriate cleanup.

- an item with an existing catkey has its metadata updated

New functionality as data changes in purl were never leading to an 856 change when needed. Since a new 856 is regenerated every time a released item changes, those changes will propagate automatically.

The volume of 856 updates is noted as a possible concern. The steady state rate of change for released items is low, but a mass remediation could mean lots of new 856s for objects whose 856 information has not really changed. Re-checking with Darsi on any concerns here. It could be optimized if the service did a sw-solr lookup to the object's current purl-based data, though it would have to be parsed out of MARCXML from the document there for comparison.

cbeer commented:
> Since a new 856 is regenerated every time a released item changes, those changes will propagate automatically.

I talked with @dlrueda earlier today, and we're both concerned that multiple 856 log entries for a single catkey/druid combination within a single day may lead to unexpected behavior.

LynnMcRae commented:
I meet with @dlrueda today at 3:30. We think the ordering of the tags is key (last one should win) and is already taken care of based on original design last summer since duplicates were already possible before this, but need to confirm.

cbeer commented on lib/sw_indexer_engine.rb:
```diff
@@ -9,17 +9,23 @@ class SwIndexerEngine < BaseIndexer::MainIndexerEngine
     targets.each do |target_key, target_value|
       next unless target_value == true && !Settings.SKIP_CATKEY_CHECK.include?(target_key)
-      targets[target_key] = false if purl_model(druid).catkey.present?
+      if purl_model(druid).catkey.present?
```
Instead of pulling out `update_marc_record_needed`, can you pull out `purl_model(druid)`? It seems like it doesn't depend on the `target` values, right?

LynnMcRae commented:
Met with @dlrueda. There is a problem with the current handling of more than one 856 in the same day in that a data change would cause 2 PURLs for the same druid to be injected in to the Symphony record if any of the data changed. This is an unlikely scenario with initial release scenarios, though possible if data changed and was re-released the same day. The chances for error increase somewhat if multiple changes affecting catkey-indexed information get published in one day. Doing a separate ticket for a general fix to Symphony-side script to only process the last entry for a catkey/druid combo as there can only be one PURL+data for a given catkey/druid combination.

tingulfsen commented:
Has there been any resolution to the questions discussed above?

peetucket commented:
@LynnMcRae : I think the issue that @dlrueda was mentioned in above is completed on her end, right?

LynnMcRae commented:
The question above about how Symphony handles mulitple incoming entries for a PURL. This has been resolved. Darsi's script has already been modified to accept only the last PURL asserted for a given catkey.

tingulfsen commented:
I see that @cbeer requested some changes - have they been made or are they now obsolete?

peetucket commented:
Chris had general concerns with the approach overall which we discussed when the work was in progress (which I believe still remain, but which Lynn has indicated we will move ahead with).  I also believe I have addressed the specific concerns with lines of code that Chris pointed out (some of them may have been hidden since I squashed commits at one point).

jmartin-sul commented:
i would also be slightly in favor of squashing the first four commits into one, since the 3rd and 4th sort of undo things from the first two.  but wouldn't hold up merging over that if it's a pain.

peetucket commented:
commits squashed down to 3
